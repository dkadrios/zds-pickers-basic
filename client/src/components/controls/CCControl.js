import { useCallback } from 'react'
import PropTypes from 'prop-types'
import { LatchPicker, PolarityPicker } from 'zds-pickers'
import makeStyles from '@material-ui/core/styles/makeStyles'
import { border, margin, padding } from 'polished'
import { useDispatch } from 'react-redux'
import Led from '../Led'
import { actions } from '../../reducers/inputControls'
import { pick } from '../../core/fp/objects'
import CCPicker from './CCPicker'
import ChannelPicker from './ChannelPicker'

const useStyles = makeStyles(({ mixins: { importantPx }, palette }) => ({
  root: {
    ...margin(0, 10),
    ...padding(10, 5, 10),
    ...border(4, 'solid', palette.border),
    borderRadius: 15,
    display: 'flex',
    flexFlow: 'column nowrap',
    justifyContent: 'space-between',
    alignItems: 'stretch',
    alignContent: 'center',

    '& figure': {
      // LED Container
      padding: 0,
      ...margin(0, 0, 10),
      '& header': {
        fontWeight: 'bold',
        fontSize: 18,
      },
    },
  },
  picker: {
    minWidth: 179,
    '& input': {
      backgroundColor: palette.accentLight, // 'lighten($color-shifter-red, 5)',
      color: palette.text.inverted,
      ...padding(4, 6),
      borderRadius: 5,
    },
    '& ul': {
      maxHeight: importantPx(160),
    },
  },
}), { name: 'CCControl' })

const CCControl = (props) => {
  const dispatch = useDispatch()
  const {
    // changeInputControlChannel,
    // changeInputControlLatching,
    // changeInputControlPolarity,
    // changeInputControlValue,
    controlId,
    lit,
    ...rest
  } = props

  const handleChangeValue = useCallback((value) => {
    dispatch(actions.changeInputControlValue(controlId, value))
  }, [controlId, dispatch])

  const handleChangeChannel = useCallback((value) => {
    dispatch(actions.changeInputControlChannel(controlId, value))
  }, [controlId, dispatch])

  const handleChangeLatching = useCallback((value) => {
    dispatch(actions.changeInputControlLatching(controlId, value))
  }, [controlId, dispatch])

  const handleChangePolarity = useCallback((value) => {
    dispatch(actions.changeInputControlPolarity(controlId, value))
  }, [controlId, dispatch])

  const classes = useStyles()

  const pickerProps = {
    ...pick(['channel', 'className', 'latching', 'polarity', 'value'])(props),
    className: classes.picker,
  }

  return (
    <div className={classes.root}>
      <Led
        label={`FS${controlId + 1}`}
        lit={lit}
        {...rest}
      />
      <CCPicker
        {...pickerProps}
        onChange={handleChangeValue}
      />
      <ChannelPicker
        {...pickerProps}
        onChange={handleChangeChannel}
      />
      <LatchPicker
        {...pickerProps}
        onChange={handleChangeLatching}
      />
      <PolarityPicker
        {...pickerProps}
        labelOff={pickerProps.latching ? 'Initially Off' : 'Normally Off'}
        labelOn={pickerProps.latching ? 'Initially On' : 'Normally On'}
        onChange={handleChangePolarity}
      />
    </div>
  )
}

CCControl.propTypes = {
  lit: PropTypes.bool.isRequired,
  controlId: PropTypes.number.isRequired,
  value: PropTypes.number.isRequired,
  latching: PropTypes.oneOf([0, 1]).isRequired,
  polarity: PropTypes.oneOf([0, 1]).isRequired,
}

export default CCControl
