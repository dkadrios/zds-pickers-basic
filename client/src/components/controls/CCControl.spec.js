import React from 'react'
import set from 'lodash/fp/set'
import flow from 'lodash/fp/flow'
import CCControl from './CCControl'

const changeInputControlChannel = jest.fn()
const changeInputControlValue = jest.fn()
const changeInputControlLatching = jest.fn()
const changeInputControlPolarity = jest.fn()

const controlId = 3

const props = {
  lit: false,
  controlId,
  changeInputControlChannel,
  changeInputControlValue,
  changeInputControlLatching,
  changeInputControlPolarity,
  disabled: false,
  value: 12,
  latching: 1,
  polarity: 0,
}

describe('CCControl tests', () => {
  beforeEach(jest.resetAllMocks)

  it('renders correctly', () => {
    shallowExpect(CCControl)(props).toMatchSnapshot()
  })

  it('renders alternate mode correctly', () => {
    shallowExpect(CCControl)(flow(
      set('latching', 0),
      set('polarity', 1),
    )(props)).toMatchSnapshot()
  })

  it('responds to CC changes', () => {
    const wrapper = shallow(<CCControl {...props} />)
    wrapper.find('WithStyles(CCPicker)').simulate('change', 13)
    expect(changeInputControlValue).toHaveBeenCalledWith(controlId, 13)
  })

  it('responds to channel changes', () => {
    const wrapper = shallow(<CCControl {...props} />)
    wrapper.find('WithStyles(ChannelPicker)').simulate('change', 9)
    expect(changeInputControlChannel).toHaveBeenCalledWith(controlId, 9)
  })

  it('responds to latching changes', () => {
    const wrapper = shallow(<CCControl {...props} />)
    wrapper.find('LatchPicker').simulate('change', 0)
    expect(changeInputControlLatching).toHaveBeenCalledWith(controlId, 0)
  })

  it('responds to polarity changes', () => {
    const wrapper = shallow(<CCControl {...props} />)
    wrapper.find('PolarityPicker').simulate('change', 1)
    expect(changeInputControlPolarity).toHaveBeenCalledWith(controlId, 1)
  })
})
