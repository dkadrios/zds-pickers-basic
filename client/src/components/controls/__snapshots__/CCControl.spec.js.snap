// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CCControl tests renders alternate mode correctly 1`] = `
<div>
  <Led
    disabled={false}
    isMidiActivity={false}
    label="FS4"
    latching={0}
    layout="left"
    lit={false}
    polarity={1}
    value={12}
  />
  <WithStyles(CCPicker)
    latching={0}
    onChange={[Function]}
    polarity={1}
    value={12}
  />
  <WithStyles(ChannelPicker)
    latching={0}
    onChange={[Function]}
    polarity={1}
    value={12}
  />
  <LatchPicker
    latching={0}
    onChange={[Function]}
    polarity={1}
    value={12}
  />
  <PolarityPicker
    labelOff="Normally Off"
    labelOn="Normally On"
    latching={0}
    onChange={[Function]}
    polarity={1}
    value={12}
  />
</div>
`;

exports[`CCControl tests renders correctly 1`] = `
<div>
  <Led
    disabled={false}
    isMidiActivity={false}
    label="FS4"
    latching={1}
    layout="left"
    lit={false}
    polarity={0}
    value={12}
  />
  <WithStyles(CCPicker)
    latching={1}
    onChange={[Function]}
    polarity={0}
    value={12}
  />
  <WithStyles(ChannelPicker)
    latching={1}
    onChange={[Function]}
    polarity={0}
    value={12}
  />
  <LatchPicker
    latching={1}
    onChange={[Function]}
    polarity={0}
    value={12}
  />
  <PolarityPicker
    labelOff="Initially Off"
    labelOn="Initially On"
    latching={1}
    onChange={[Function]}
    polarity={0}
    value={12}
  />
</div>
`;
